beacon_node/store/src/hot_cold_store.rs

```diff
@@ -3361,20 +3378,106 @@ fn no_state_root_iter() -> Option<std::iter::Empty<Result<(Hash256, Slot), Error
-        .find(|result| match result {
-            Ok((_, result_slot)) => *result_slot == target_slot,
-            Err(_) => true, // Keep errors intact and stop on the first occurrence
-        })
-        .transpose()
-        .map(|opt| opt.map(|(root, _)| root))
+) -> Result<Hash256, StateSummaryIteratorError> {
+    // Use the state itself for recent roots
+    if let Ok(target_state_root) = from_state.get_state_root(target_slot) {
+        return Ok(*target_state_root);
+    }
+
+    let mut state_root = {
+        // We can not start loading summaries from `state_root` since its summary has not yet been
+        // imported. This code path is called during block import.
+        //
+        // We need to choose a state_root to start that is
+        // - An ancestor of `from_state`, AND
+        // - Its state summary is already written (and not pruned) in the DB
+        // - Its slot is >= target_slot
+        //
+        // If we get to this codepath, (target_slot not in state's state_roots) it means that
+        // `state.slot()` is greater than `SlotsPerHistoricalRoot`, and `target_slot < state.slot()
+        // - SlotsPerHistoricalRoot`.
+        //
+        // Values we could start from:
+        // - `state.slot() - 1`: TODO if we don't immediately commit all each state to the DB
+        //   individually, we may be attempting to read a state summary that is stored in a DB ops
+        //   vector but not yet written to the DB. Also starting from this slot is wasteful as we
+        //   know that the target slot is `< state.slot() - SlotsPerHistoricalRoot`.
+        // - `state.slot() - SlotsPerHistoricalRoot`: The most efficient slot to start. But we risk
+        //   jumping to a state summary that has already been pruned. See the `max(.., split_slot)`
+        //   below
+        let oldest_slot_in_state_roots = from_state
+            .slot()
+            .saturating_sub(Slot::new(E::SlotsPerHistoricalRoot::to_u64()));
+
+        // Don't start with a slot that prior to the finalized state slot. We may be attempting to read
+        // a hot state summary that has already been pruned as part of the migration and error. HDiffs
+        // can reference diffs with a slot prior to the finalized checkpoint. But those are sparse so
+        // the probabiliy of hitting `MissingSummary` error is high. Instead, the summary for the
+        // finalized state is always available.
+        let start_slot = std::cmp::max(oldest_slot_in_state_roots, store.get_split_slot());
+
+        *from_state
+            .get_state_root(start_slot)
+            .map_err(|_| StateSummaryIteratorError::OutOfBoundsInitialSlot)?
+    };
+
+    let mut previous_slot = None;
+
+    loop {
+        let state_summary = store
+            .load_hot_state_summary(&state_root)
+            .map_err(|e| StateSummaryIteratorError::LoadSummaryError(Box::new(e)))?
+            .ok_or(StateSummaryIteratorError::MissingSummary(state_root))?;
```

Comment:
This could fail if `state_root` is pruned between when we took the `split` lock. We should probably hold the split lock from the start of this function (after the lookup in the state misses). This should be OK for performance because loading summaries is cheap and we're never iterating that far back.